cranelift/codegen/src/simple_preopt.rs

```diff
@@ -1,796 +0,0 @@
-                                    pos.func
-                                        .dfg
-                                        .replace(inst)
-                                        .BinaryImm64(opcode, ty, new_imm, new_arg);
-                                    imm = new_imm;
-                                    arg = new_arg;
-                                }
-                            }
-                        }
-                    }
-
-                    Opcode::UshrImm | Opcode::SshrImm => {
-                        if pos.func.dfg.ctrl_typevar(inst).bytes() <= native_word_width
-                            && try_fold_extended_move(pos, inst, opcode, arg, imm)
-                        {
-                            return;
-                        }
-                    }
-
-                    _ => {}
-                };
-
-                // Replace operations that are no-ops.
-                match (opcode, imm.into(), ty) {
-                    (Opcode::IaddImm, 0, _)
-                    | (Opcode::ImulImm, 1, _)
-                    | (Opcode::SdivImm, 1, _)
-                    | (Opcode::UdivImm, 1, _)
-                    | (Opcode::BorImm, 0, _)
-                    | (Opcode::BandImm, -1, _)
-                    | (Opcode::BxorImm, 0, _)
-                    | (Opcode::RotlImm, 0, _)
-                    | (Opcode::RotrImm, 0, _)
-                    | (Opcode::IshlImm, 0, _)
-                    | (Opcode::UshrImm, 0, _)
-                    | (Opcode::SshrImm, 0, _) => {
-                        // Alias the result value with the original argument.
-                        replace_single_result_with_alias(&mut pos.func.dfg, inst, arg);
-                    }
-                    (Opcode::ImulImm, 0, ty) | (Opcode::BandImm, 0, ty) if ty != I128 => {
-                        // Replace by zero.
-                        pos.func.dfg.replace(inst).iconst(ty, 0);
-                    }
-                    (Opcode::BorImm, -1, ty) if ty != I128 => {
-                        // Replace by minus one.
-                        pos.func.dfg.replace(inst).iconst(ty, -1);
-                    }
-                    _ => {}
-                }
-            }
-
-            InstructionData::IntCompare { opcode, cond, args } => {
-                debug_assert_eq!(opcode, Opcode::Icmp);
-                if let Some(imm) = resolve_imm64_value(&pos.func.dfg, args[1]) {
-                    if pos.func.dfg.ctrl_typevar(inst).bytes() <= native_word_width {
-                        pos.func.dfg.replace(inst).icmp_imm(cond, args[0], imm);
-                    }
-                }
-            }
-
-            _ => {}
-        }
-    }
-
-    /// Fold comparisons into branch operations when possible.
-    ///
-    /// This matches against operations which compare against zero, then use the
-    /// result in a conditional branch.
-    fn branch_opt(pos: &mut FuncCursor, inst: Inst) {
```

Comment:
I think we're still unable to port this optimization to egraphs. That shouldn't block this PR, but it would be nice to optimize control flow in the future as well :)

Reply:
Specifically, that issue is tracked in #6106.